### ISSUE TITLE
PBClientFactory.getPerspective does not handle errors in authentication

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7606]
|<img alt="syver's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| syver reported|
|-|-|
|Trac ID|trac#394|
|Type|enhancement|
|Created|2003-11-12 18:28:28Z|

<details><summary>Searchable metadata</summary>

```
trac-id__394 394
type__enhancement enhancement
reporter__syver syver
priority__low low
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__invalid invalid
component__pb pb
keywords__ 
time__1068661708000000 1068661708000000
changetime__1469226047623590 1469226047623590
version__ 
owner__ 
cc__exarkun cc__itamarst cc__syver
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="syver's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|syver commented|
|-|-|

```
#!html
<pre>
The code from twisted 1.1.0 is as follows
    def _cbGetPerspective(self, identityWrapper, serviceName, perspectiveName, 
client):
        return identityWrapper.callRemote(
            "attach", serviceName, perspectiveName, client)

When the user is not authorized the identityWrapper argument is None. This 
causes an AttributeError when attempting to send callRemote to the None object.

I stole the following code from pb.IdentityConnector and it seems to do the 
right thing.

    def _cbGetPerspective(self, identityWrapper, serviceName, perspectiveName, 
client):
        if not identityWrapper:
            from twisted import cred
            e = cred.error.Unauthorized("invalid username or password")
            return defer.fail(e)
        return identityWrapper.callRemote(
            "attach", serviceName, perspectiveName, client)

I am aware that getPerspective is deprecated and therefore there won't be any 
priority in fixing it. It could be nice if anyone else has the same problem and 
can recognize it here in the bug database
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
Tried patch, it failed my test, don't care enough to spend any more time on it.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
Perspectives are so old and sad that no one is going to resolve this.  Rejecting.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|GitHub commented|
|-|-|

In changeset 4f0baf8aa1c73f9dcde4f6acab67b57f57a4cae7
```
#!CommitTicketReference repository="" revision="4f0baf8aa1c73f9dcde4f6acab67b57f57a4cae7"
Merge pull request #394 from twisted/8654-einarfd-ssh-userauth-2

Author: einarfd
Reviewer: rodrigc
Fixes: #8654 

Port twisted.conch.ssh.userauth to Python 3
```
